From e9627c404ec2a3941e246cb7729f0694601fb18a Mon Sep 17 00:00:00 2001
From: "Pedro P. Candel" <pedro@joyent.com>
Date: Mon, 17 Oct 2016 13:40:25 +0200
Subject: [PATCH] TOOLS-1574 sdcadm experimental avail crashes due to missing
 agent image uuids Reviewed by: Trent Mick <trent.mick@joyent.com>

---
 lib/cli/do_add_new_agent_svcs.js | 90 +++++++++++++++++++++++++-------
 lib/cli/do_update_other.js       | 52 +++++++++++++++++-
 2 files changed, 123 insertions(+), 19 deletions(-)

diff --git a/lib/cli/do_add_new_agent_svcs.js b/lib/cli/do_add_new_agent_svcs.js
index f22c900..62e782c 100644
--- a/lib/cli/do_add_new_agent_svcs.js
+++ b/lib/cli/do_add_new_agent_svcs.js
@@ -5,8 +5,11 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
+
+var fs = require('fs');
+var util = require('util');
 var vasync = require('vasync');
 
 var errors = require('../errors');
@@ -59,7 +62,38 @@ function do_add_new_agent_svcs(subcmd, opts, args, cb) {
         agentServices[n].metadata[logLevelKey] = 'info';
     });
 
+
+    // The first time we add agent services to SAPI we'll use the HN image
+    // version to create the service, assuming that's the version installed
+    // everywhere across the whole SDC setup
+    function getAgentImages(callback) {
+        vasync.forEachPipeline({
+            func: function (agent, next) {
+                var name = agent.name;
+                var imgUUIDPath = util.format(
+                    '/opt/smartdc/agents/lib/node_modules/%s/image_uuid',
+                    name);
+                fs.readFile(imgUUIDPath, {
+                    encoding: 'utf8'
+                }, function (err, data) {
+                    if (err) {
+                        self.sdcadm.log.error({err: err},
+                            'Error reading agent image uuid');
+                        next();
+                        return;
+                    }
+                    agentServices[name].params.image_uuid = data.trim();
+                    next();
+                });
+            },
+            inputs: agentNames.map(function (agent) {
+                return agentServices[agent];
+            })
+        }, callback);
+    }
+
     var newAgentServices = [];
+    var updateAgentServices = [];
     // Used by history:
     var history;
     var changes = [];
@@ -113,17 +147,21 @@ function do_add_new_agent_svcs(subcmd, opts, args, cb) {
                         if (!svcs.length) {
                             newAgentServices.push(agent);
                         }
+                        if (!svcs[0].params.image_uuid) {
+                            agentServices[agent] = svcs[0];
+                            updateAgentServices.push(agent);
+                        }
                         return callback();
                     });
                 },
                 inputs: Object.keys(agentServices)
-            }, function (err) {
-                if (err) {
-                    return next(err);
-                }
-                return next();
-            });
+            }, next);
+        },
+
+        function getAgentImgVersions(_, next) {
+            getAgentImages(next);
         },
+
         function saveChangesToHistory(_, next) {
             newAgentServices.forEach(function (s) {
                 changes.push({
@@ -133,8 +171,18 @@ function do_add_new_agent_svcs(subcmd, opts, args, cb) {
                     },
                     type: 'create-service'
                 });
+            });
 
+            updateAgentServices.forEach(function (s) {
+                changes.push({
+                    service: {
+                        name: s,
+                        type: 'agent'
+                    },
+                    type: 'update-service'
+                });
             });
+
             self.sdcadm.history.saveHistory({
                 changes: changes
             }, function (err, hst) {
@@ -155,19 +203,25 @@ function do_add_new_agent_svcs(subcmd, opts, args, cb) {
                         params: agentServices[agent]
                     }, 'Adding new agent service');
                     self.sdcadm.sapi.createService(agent, app.uuid,
-                        agentServices[agent], function (err) {
-                            if (err) {
-                                return callback(err);
-                            }
-                            return callback();
-                    });
+                        agentServices[agent], callback);
                 }
-            }, function (err) {
-                if (err) {
-                    return next(err);
+            }, next);
+        },
+
+        function updateAgentsServices(_, next) {
+            vasync.forEachParallel({
+                inputs: updateAgentServices,
+                func: function updateAgentSvc(agent, callback) {
+                    progress('Updating service for agent \'%s\'', agent);
+                    self.log.trace({
+                        service: agent,
+                        params: agentServices[agent]
+                    }, 'Updating agent service');
+                    self.sdcadm.sapi.updateService(agentServices[agent].uuid, {
+                        params: agentServices[agent].params
+                    }, callback);
                 }
-                return next();
-            });
+            }, next);
         }
     ]}, function (err) {
         progress('Add new agent services finished (elapsed %ds).',
diff --git a/lib/cli/do_update_other.js b/lib/cli/do_update_other.js
index d8e696c..f027bcf 100644
--- a/lib/cli/do_update_other.js
+++ b/lib/cli/do_update_other.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright 2016 Joyent, Inc.
+ * Copyright 2017 Joyent, Inc.
  */
 
 var util = require('util');
@@ -700,6 +700,56 @@ function do_update_other(subcmd, opts, args, cb) {
                 }
                 return next();
             });
+        },
+
+        function updateAgentsImages(ctx, next) {
+            var svcsToUpdate = [];
+            ctx.svcs.forEach(function (svc) {
+                if (svc.type === 'agent' && svc.params &&
+                        !svc.params.image_uuid &&
+                        svc.name !== 'dockerlogger') {
+                    svcsToUpdate.push(svc);
+                }
+            });
+
+            function updateAgentImage(agent, callback) {
+                vasync.pipeline({
+                    funcs: [
+                        function readAgentImg(_, _cb) {
+                            var name = agent.name;
+                            var imgUUIDPath = util.format(
+                                '/opt/smartdc/agents/lib/' +
+                                'node_modules/%s/image_uuid',
+                                name);
+                            fs.readFile(imgUUIDPath, {
+                                encoding: 'utf8'
+                            }, function (err, data) {
+                                if (err) {
+                                    self.sdcadm.log.error({err: err},
+                                        'Error reading agent image uuid');
+                                    _cb(err);
+                                    return;
+                                }
+                                agent.params.image_uuid = data.trim();
+                                _cb();
+                            });
+
+                        },
+                        function updateAgentImg(_, _cb) {
+                            progress('Updating service for agent \'%s\'',
+                                    agent.name);
+                            updateService(agent.uuid, {
+                                params: agent.params
+                            }, _cb);
+                        }
+                    ]
+                }, callback);
+            }
+
+            vasync.forEachParallel({
+                inputs: svcsToUpdate,
+                func: updateAgentImage
+            }, next);
         }
     ]}, function (err) {
         if (!history) {
-- 
2.21.0

